nebula_decoders/include/nebula_decoders/nebula_decoders_robosense/decoders/robosense_decoder.hpp

```diff
@@ -0,0 +1,246 @@
+
+  /// @brief The last decoded packet
+  typename SensorT::packet_t packet_;
+  /// @brief The last azimuth processed
+  int last_phase_;
+  /// @brief The timestamp of the last completed scan in nanoseconds
+  uint64_t output_scan_timestamp_ns_;
+  /// @brief The timestamp of the scan currently in progress
+  uint64_t decode_scan_timestamp_ns_;
+  /// @brief Whether a full scan has been processed
+  bool has_scanned_;
+
+  rclcpp::Logger logger_;
+
+  /// @brief Validates and parse PandarPacket. Currently only checks size, not checksums etc.
+  /// @param pandar_packet The incoming PandarPacket
+  /// @return Whether the packet was parsed successfully
+  bool parsePacket(const pandar_msgs::msg::PandarPacket & pandar_packet)
+  {
+    if (pandar_packet.size < sizeof(typename SensorT::packet_t)) {
+      RCLCPP_ERROR_STREAM(
+        logger_, "Packet size mismatch:" << pandar_packet.size << " | Expected at least:"
+                                         << sizeof(typename SensorT::packet_t));
+      return false;
+    }
+    if (std::memcpy(&packet_, pandar_packet.data.data(), sizeof(typename SensorT::packet_t))) {
+      return true;
+    }
+
+    RCLCPP_ERROR(logger_, "Packet memcopy failed");
+    return false;
+  }
+
+  /// @brief Converts a group of returns (i.e. 1 for single return, 2 for dual return, etc.) to
+  /// points and appends them to the point cloud
+  /// @param start_block_id The first block in the group of returns
+  /// @param n_blocks The number of returns in the group
+  void convertReturns(size_t start_block_id, size_t n_blocks)
+  {
+    uint64_t packet_timestamp_ns = robosense_packet::get_timestamp_ns(packet_);
+    uint32_t raw_azimuth = packet_.body.blocks[start_block_id].get_azimuth();
+
+    std::vector<const typename SensorT::packet_t::body_t::block_t::unit_t *> return_units;
+
+    for (size_t channel_id = 0; channel_id < SensorT::packet_t::N_CHANNELS; ++channel_id) {
+      // Find the units corresponding to the same return group as the current one.
+      // These are used to find duplicates in multi-return mode.
+      return_units.clear();
+      for (size_t block_offset = 0; block_offset < n_blocks; ++block_offset) {
+        return_units.push_back(
+          &packet_.body.blocks[block_offset + start_block_id].units[channel_id]);
+      }
+
+      for (size_t block_offset = 0; block_offset < n_blocks; ++block_offset) {
+        auto & unit = *return_units[block_offset];
+
+        if (unit.distance.value() == 0) {
+          continue;
+        }
+
+        auto distance = getDistance(unit);
+
+        if (distance < SensorT::MIN_RANGE || distance > SensorT::MAX_RANGE) {
+          continue;
+        }
+
+        // Keep only last (if any) of multiple points that are too close
+        if (block_offset != n_blocks - 1) {
+          for (size_t return_idx = 0; return_idx < n_blocks; ++return_idx) {
+            if (return_idx == block_offset) {
+              continue;
+            }
+          }
+        }
```

Comment:
This loop, as it is, does nothing; It seems that the dual return distance threshold check has been removed. We should probably keep this consistent with the other sensor decoders and include it. @amc-nu @drwnz